nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -1889,6 +1904,61 @@ impl<'a> BlueprintBuilder<'a> {
     ) {
         self.pending_mgs_updates.remove(baseboard_id);
     }
+
+    fn zone_image_artifact(
+        repo: Option<&TufRepoDescription>,
+        zone_kind: ZoneKind,
+    ) -> BlueprintZoneImageSource {
+        repo.and_then(|repo| {
+            repo.artifacts
+                .iter()
+                .find(|artifact| {
+                    zone_kind.is_control_plane_zone_artifact(&artifact.id)
+                })
+                .map(BlueprintZoneImageSource::from_available_artifact)
+        })
+        .unwrap_or(BlueprintZoneImageSource::InstallDataset)
+    }
+
+    /// Try to find an artifact in either the current or previous release repo
+    /// that contains an image for a zone of the given kind; see RFD 565 §9.
+    /// Defaults to the install dataset.
+    pub(crate) fn zone_image_source(
+        &self,
+        zone_kind: ZoneKind,
+    ) -> BlueprintZoneImageSource {
+        let new_repo = self.input.tuf_repo();
+        let old_repo = self.input.old_repo();
+        let new_artifact = Self::zone_image_artifact(new_repo, zone_kind);
+        let old_artifact = Self::zone_image_artifact(old_repo, zone_kind);
+        match OrderedComponent::from(zone_kind) {
+            // Nexus can only be updated if all non-Nexus zones have been updated.
+            OrderedComponent::NexusZone => {
+                if self.sled_ids_with_zones().any(|sled_id| {
+                    self.current_sled_zones(
+                        sled_id,
+                        BlueprintZoneDisposition::is_in_service,
+                    )
+                    .filter(|z| {
+                        OrderedComponent::from(z.zone_type.kind())
+                            == OrderedComponent::NonNexusOmicronZone
+                    })
+                    .any(|z| z.image_source != new_artifact)
```

Comment:
I don't think this is the correct `new_artifact`, right? This `new_artifact` is the Nexus zone we're trying to add, but we want to check `z.zone_type`'s artifact from `new_repo`?

Reply:
Right, thanks! Also fixed in 3611b428c.